juju/juju
# addition: show space command

## Checklist


----

## Description of change

- addition of the show-space command 
- bump facade 
- tests should show cli and apiserver behavior

## QA steps
Test must run on a provider supporting spaces
- test on aws
- test on maas
- create/use existing space (here we call it db)
- add an application with endpoint bind
- deploy a machine/application with space constraint

```sh
❯ juju deploy mediawiki --constraints spaces=db
Located charm "cs:mediawiki-19".
Deploying charm "cs:mediawiki-19".
❯  juju deploy mysql --bind slave=db 
Located charm "cs:mysql-58".
Deploying charm "cs:mysql-58".
❯ juju add-machine --constraints spaces=db      
created machine 5
```

should, in turn, return the following:
```
❯ juju show-space db                
space:
  id: "2"
  name: db
  subnets:
  - cidr: 172.31.16.0/20
    provider-id: subnet-367aa55c
    provider-network-id: vpc-a12addcb
    vlan-tag: 0
    zones:
    - eu-central-1a
  - cidr: 252.16.0.0/12
    provider-id: subnet-367aa55c-INFAN-172-31-16-0-20
    provider-network-id: vpc-a12addcb
    vlan-tag: 0
    zones:
    - eu-central-1a
applications:
- mediawiki
- mysql
machine-count: 3
```

e.g in aws alpha is the default space, which contains every subnet by default. Every application does have a default bind to alpha. Therefore the output should, in the above case:
- contains every application:
- contains every machine count, which was explicitly deployed on alpha. As we only reason for private addresses. If we deploy a machine:
   -  it will have a private address bound to the space defined.
   -  a public address bound to `alpha`, which we don't need to reason about. That's why it will show 0
- contain all subnets, unless we allocated them to another space.
```
❯ juju show-space alpha
space:
  id: "0"
  name: alpha
  subnets:
  - cidr: 172.31.0.0/28
    provider-id: subnet-012ea464634aa2e3d
    provider-network-id: vpc-a12addcb
    vlan-tag: 0
    zones:
    - eu-central-1a
  - cidr: 252.0.0.0/20
    provider-id: subnet-012ea464634aa2e3d-INFAN-172-31-0-0-28
    provider-network-id: vpc-a12addcb
    vlan-tag: 0
    zones:
    - eu-central-1a
applications:
- <every application>
machine-count: 0 <as we do not deploy --bind explicitly on that space, every other machine lands with a second public address bound to `0`, which we do not care about.>
```

## Documentation changes

following design cli: https://docs.google.com/document/d/1tE5GMF9Uw8W7QQoybgA_vs_RqCHQdbKXmQT_Kaaj6Pk/edit#heading=h.m8owshq36rsr
